app/src/main/java/ch/epfl/sdp/drone3d/map/offline/OfflineMapSaverImpl.kt

```diff
@@ -0,0 +1,225 @@
+import com.mapbox.mapboxsdk.maps.Style
+import com.mapbox.mapboxsdk.offline.*
+import kotlinx.serialization.Serializable
+import kotlinx.serialization.decodeFromString
+import kotlinx.serialization.encodeToString
+import kotlinx.serialization.json.Json
+import timber.log.Timber
+
+
+/**
+ * Used internally to facilitate the serialization of the OfflineRegionMetadata data class
+ */
+@Serializable
+private data class SerializableRegionMetadata(val name:String, val latSouth:Double, val lonWest:Double,
+                                              val latNorth:Double, val lonEast:Double,val tileCount:Long, val zoom:Double,
+                                              val isDownloading:Boolean)
+
+class OfflineMapSaverImpl(val context:Context, private val styleUri:String):OfflineMapSaver {
+
+    companion object{
+        private const val JSON_CHARSET = "UTF-8"
+        const val MAX_ZOOM = 20.0
+        const val MAX_ZOOM_DISCREPANCY = 6.0 //Maximum difference between max zoom and min zoom
+        const val TILE_LIMIT = 6000L //6000 tiles corresponds to Greater London with zoom 0-15
+        private val totalTileCount = MutableLiveData<Long>(0)
+        private val offlineRegions = MutableLiveData<Array<OfflineRegion>>()
+        //Stores the ids of the regions currently downloading
+        private val ongoingDownloads = mutableSetOf<Long>()
+        /**
+         * Serialize the metadata of the provided [region]
+         */
+        fun serializeMetadata(metadata: OfflineRegionMetadata):ByteArray{
+            val internalMetadata = SerializableRegionMetadata(metadata.name,metadata.bounds.latSouth,
+                metadata.bounds.lonWest,metadata.bounds.latNorth,metadata.bounds.lonEast,metadata.tileCount,
+                metadata.zoom,metadata.isDownloading)
+
+            return Json.encodeToString(internalMetadata).toByteArray(charset(JSON_CHARSET))
+        }
+
+        /**
+         * Deserialize the metadata [serialized]
+         */
+         fun deserializeMetadata(serialized: ByteArray):OfflineRegionMetadata{
+            val metadata:SerializableRegionMetadata = Json.decodeFromString(String(serialized))
+            val bounds = LatLngBounds.Builder()
+                            .include(LatLng(metadata.latSouth, metadata.lonWest))
+                            .include(LatLng(metadata.latNorth, metadata.lonEast))
+                            .build()
+
+            return OfflineRegionMetadata(metadata.name,bounds,metadata.tileCount,metadata.zoom,
+                metadata.isDownloading)
+        }
+
+        /**
+         * Returns the metadata of the region
+         */
+        fun getMetadata(region:OfflineRegion):OfflineRegionMetadata{
+            return deserializeMetadata(region.metadata)
+        }
+
+        /**
+         * Returns the centered camera position on [offlineRegion]
+         */
+        fun getRegionLocation(offlineRegion: OfflineRegion): CameraPosition {
```

Comment:
Same question for all the methode above, you use them only in this class.

Reply:
getMetadata is used by the UI to display the name of the region
I made the serialisation/deserialization public to be able to test them.
getRegionLocation is not used yet by the UI but it will be userful later on to zoom on a selected saved region

Reply:
You can annotate function to be public only for testing